text/0000-eager.md

```diff
@@ -0,0 +1,424 @@
+
+```
+macro foo {
+    () => { a::bar!(); }
+}
+
+mod a {
+    macro bar {
+        () => { let b = 42; b }
+    }
+}
+
+fn main() {
+    foo!();
+}
+```
+
+There is a 'global' scope due to the enclosing crate. The following identifiers
+have scope sets:
+
+* `foo: {in_crate, out_crate}`
+* `a: {in_crate, out_crate}`
+* `bar: {in_crate, out_crate, in_a, out_a}`
+* `main: {in_crate, out_crate}`
+* `foo!: {in_crate, out_crate, in_main, out_main}`
+
+For the macro definitions, the pending set is given by the set of scopes on the
+macro name.
+
+Looking up the binding for `foo!` is trivial. After one step of expansion we get
+
+```
+fn main() {
+    a::bar!();
+}
+```
+
+`a` has scope set `{in_crate, out_crate, intro_foo1, in_main}`, so looking it up is
+straightforward. We then use `a`'s scopes to look up bar.
+
+After the next step of expansion we get
+
+```
+fn main() {
+    let b = 42;
+    b
+}
+```
+
+Both uses of the name `b` have the same set of scopes:
+`{in_crate, out_crate, in_a, out_a, intro_bar, in_main, in_b, out_b}`.
+
+For the eager expansion case, there is a step before expansion - we must find
+the macro to expand. Since we haven't parsed the enclosing macro, we can't use
+the hygiene information from the eager macro use site. That means we look up the
+eager macro in the context of the enclosing macro.
+
+When we expand the eager use, we still use the pending set of scopes for the
+macro being expanded, but not for the enclosing macro. We also add introduction
+scopes for the macro being expanded, but not the enclosing macro. In these ways
+eager expansion mimics regular expansion (i.e., the resulting sets of scopes are
+similar). When we parse, we add scopes as normal. Note that since identifiers
+are expanded before parsing, they will get both inside and outside edge
+scopes (c.f., regular expansion). Finally, because we can't parse the eagerly
+expanded macro until we have expanded the enclosing macro, we must keep the
+pending scope set for the eagerly expanded macro around longer than usual.
+Example:
+
+```
+macro foo {
+    () => { $!a::bar(); }
+}
+
+mod a {
+    macro bar {
+        () => { let b = 42; b }
+    }
+}
+
+fn main() {
+    foo!();
+}
+```
+
+The starting scope sets are the same as before. The first expansion step is to
+expand `$!a::bar()`. We get something like
+
+```
+macro foo {
+    () => { let b = 42; b }
+}
+```
+
+Where the `b`s have the `intro_bar` scope and we keep the pending scopes for
+`bar` around. Then we expand `foo!()`, we get
+
+```
+fn main() {
+    let b = 42;
+    b
+}
+```
+
+with `b: {in_crate, out_crate, in_a, out_a, intro_bar, in_main, in_b, out_b}`.
+The final scope set is the same as for regular expansion.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Adds another moving part to the macro system.
+
+Since we try to keep behaviour as close as possible to regular expansion, it may
+not be obvious when to use either flavour.
+
+
+# Alternatives
```

Comment:
Alternative: macro-level $let which lets you do things like `$$let $x = concat_idents(...); foo!($x)`. This does the same thing as eager expansion, but is more readable.